yodo1/rabbitmq/http_client.py

```diff
@@ -0,0 +1,270 @@
+import httpx
+import pika
+from tenacity import AsyncRetrying, Retrying, stop_after_attempt, wait_random
+
+logger = logging.getLogger("yodo1.rabbitmq")
+
+
+class MQSendFailedException(Exception):
+    pass
+
+
+class RabbitHttpSender:
+    def __init__(
+        self,
+        uri: str,
+        *,
+        global_max_retry: int = 3,
+        apm_client: elasticapm.Client = None,
+    ):
+        self.uri = uri
+
+        uri_obj = urlparse(uri)
+        self.host = uri_obj.hostname
+        self.username = uri_obj.username
+        self.password = uri_obj.password
+        self.virtual_host = uri_obj.path[1:]
+        self.global_max_retry = global_max_retry
+        self.apm_client: elasticapm.Client = apm_client
+
+        if self.apm_client:
+            elasticapm.instrument()
+
+    def declare_exchange(self, exchange_name: str) -> None:
+        """
+        Declare exchange on start up. We should declare exchange in the code.
+        We use pika here for convinience,
+        :param queue_name:
+        :param exchange_name:
+        :return:
+        """
+        uri = f"https://{self.username}:{self.password}@{self.host}/{self.virtual_host}"
+        connect_params = pika.URLParameters(uri)
+        connection = pika.BlockingConnection(connect_params)
+        channel = connection.channel()
+
+        channel.exchange_declare(
+            exchange=exchange_name, exchange_type="fanout", durable=True
+        )
+        logger.info(f"Exchange {exchange_name} declared")
+        channel.close()
+        connection.close()
+
+    def publish(
+        self,
+        *,
+        exchange_name: str,
+        message_body: Dict,
+        event_name: str = None,
+        properties: Dict = None,
+        routing_key: str = "",
+        max_retry: int = None,
+        max_delay_on_retry: int = 30,
+    ) -> None:
+        """
+        Publish message using sync http request
+        :param event_name: key event name like `app-release, user-register`
+        :param exchange_name: target exchange name
+        :param message_body: message body, should be able to call json.dumps
+        :param properties: message properties
+        :param routing_key: routing key
+        :param max_retry: max retry count, will use global_max_retry when it set as empty
+        :param max_delay_on_retry: max delay when retrying to send message
+        :return:
+        """
+        # setup trace context
+        traceparent_string = self._start_trace(event_name=event_name)
+
+        if max_retry is None:
+            max_retry = self.global_max_retry
+
+        # Prepare message and url
+        message = self._build_message_json(
+            message_body=message_body,
+            properties=properties,
+            routing_key=routing_key,
+            traceparent_string=traceparent_string,
+            event_name=event_name,
+        )
+        url = f"https://{self.host}/api/exchanges/{quote_plus(self.virtual_host)}/{exchange_name}/publish"
```

Comment:
Maybe we need wrapper these code into a function.

Reply:
It is already highly wrapped as function `_start_trace` and  `_build_message_json`. I think it is clear enough. Otherwise we will have way to many function. And when we need to do another change it will increase the params count.

Reply:
Sorry for the misexpression. 
I want to say that the `duplicate code` in function `async_publish` and `publish`,  that's  not important. 
Some code obsessives. hhhh :)